c/uca/sifter/unisifeggs.cpp

```diff
@@ -0,0 +1,440 @@
+        auto const it = natural_numeric_value_.find(c);
+        if (it == natural_numeric_value_.end()) {
+            return std::nullopt;
+        } else {
+            return it->second;
+        }
+    }
+
+    CodePointSet const& CodePointsWithNumericValue() const {
+        return code_points_with_numeric_value_;
+    }
+
+   private:
+    void ReadMultiPropertyFile(std::filesystem::path const& file) {
+        std::ifstream prop_list(file);
+        CHECK(prop_list.good());
+        std::cout << "Reading " << file << "...\n";
+        for (std::string line; std::getline(prop_list, line);) {
+            // See https://www.unicode.org/reports/tr44/#UnicodeData.txt.
+            auto const record = strip_comment(line);
+            if (record.empty()) {
+                continue;
+            }
+            switch (std::count(record.begin(), record.end(), ';')) {
+                case 1: {
+                    auto const [cp, property_name] = fields<2>(record);
+                    CodePointRange const range = parse_hex_codepoint_range(cp);
+                    binary_properties_[std::string(property_name)].Include(
+                        range);
+                    break;
+                }
+                case 2:  // Enumerated property.
+                    break;
+                default:
+                    std::terminate();
+            }
+        }
+    }
+
+    // Keys are C, L, M, N, P, S, Z.
+    std::map<char, CodePointSet> coarse_general_category_;
+    // Keyed on short alias.
+    std::map<std::string, CodePointSet, std::less<>> general_category_;
+    std::map<std::string, CodePointSet, std::less<>> binary_properties_;
+    // Default <code point> omitted.
+    std::map<char32_t, char32_t> simple_uppercase_mapping_;
+    // Default <code point> omitted.
+    std::map<char32_t, char32_t> simple_lowercase_mapping_;
+    // Default 0 omitted.
+    std::map<char32_t, std::uint8_t> canonical_combining_class_;
+    std::map<char32_t, std::uint64_t> natural_numeric_value_;
+    // Characters with a non-NaN Numeric_Value.
+    CodePointSet code_points_with_numeric_value_;
+};
+
+typedef unsigned short UShort16;
+
+typedef unsigned int UInt32;
+
+UCD* const ucd = new UCD();
+
+static bool isSifterNonSpacingCombining(char32_t c) {
+    // The DUCET Sifter is using these PUA code points in artificial
+    // decompositions. They are treated like gc=Mn, so that they get secondary
+    // weights.
+    return 0xF8F0 <= c && c <= 0xF8FF;
+}
+
+static bool isSpecialCJKIdeograph(char32_t c) {
+    return 0xFA0E <= c && c <= 0xFA29 &&
+           (c <= 0xFA0F || c == 0xFA11 || c == 0xFA13 || c == 0xFA14 ||
+            c == 0xFA1F || c == 0xFA21 || c == 0xFA23 || c == 0xFA24 ||
+            0xFA27 <= c);
+}
+
+extern "C" int unisift_IsAlphabetic(UInt32 c) {
+    return ucd->BinaryPropertySet("Alphabetic").contains(c) &&
+           !isSpecialCJKIdeograph(c) &&
+           !(c == 0x1E13C || c == 0x1E13D || c == 0x16B42 || c == 0x16B43);
```

Comment:
```suggestion
    return ucd->BinaryPropertySet("Alphabetic").contains(c) &&
            // HACK: TODO: Deal with the 12 special CJK ideographs
            // ("compatibility" but no decompositions) in sift() before
            // testing Alphabetic.
            !isSpecialCJKIdeograph(c) &&
            // Ken: Exceptions for gc=Lm extenders to match his sifter.
            // TODO: Look to modify this behavior for gc=Lm extenders in the future.
            !(c == 0x1E13C || c == 0x1E13D || c == 0x16B42 || c == 0x16B43);
```
(copied from unisifex.c)